ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -384,31 +384,43 @@ private String getJoinFieldName(
         final String sourceAlias,
         final LogicalSchema sourceSchema
     ) {
-      Optional<String> joinField = getJoinFieldFromExpr(
+      Optional<String> joinFieldName = getJoinFieldNameFromExpr(
           comparisonExpression.getLeft(),
           sourceAlias,
           sourceSchema
       );
 
-      if (!joinField.isPresent()) {
-        joinField = getJoinFieldFromExpr(
+      if (!joinFieldName.isPresent()) {
+        joinFieldName = getJoinFieldNameFromExpr(
             comparisonExpression.getRight(),
             sourceAlias,
             sourceSchema
         );
       }
 
-      return joinField
-          .orElseThrow(() -> new KsqlException(
-              String.format(
-                  "%s : Invalid join criteria %s. Could not find a join criteria operand for %s. ",
-                  comparisonExpression.getLocation().map(Objects::toString).orElse(""),
-                  comparisonExpression, sourceAlias
-              )
-          ));
+      if (!joinFieldName.isPresent()) {
+        // Should never happen
+        throw new IllegalStateException("Cannot find join field name");
```

Comment:
I think this will happen!

As part of this change it would be good to improve the test coverage in QTT, (`QueryTranslationTest`, i.e. the one that used the JSON based files), of join criteria. See `joins.json`.

This about what other things a user may add to the join criteria, e.g. they could use UDFs on either side, or UDAFs. So there's four extra tests - all of which will, I think, hit this `if` branch.

See if you can come up with any other ways a user may misuse the `ON` criteria, e.g. how about a literal?  and add suitable tests to cover, and ensure the error message is helpful.